File: a29/a29.py
#! /usr/bin/python3

print("Autors: Krišjānis Petručeņa kp22084")
print("Uzdevums: A29 - izvadīt divu naturālu skaitļu kopīgos dalītājus")
print("Programma izveidota: 13.09.2022")
print()

# "high-level" loģika uzdevuma izpildei
def main():
    proceed = True
    while proceed == True:
        a, b = query_two_naturals()
        result = get_common_divisors(a,b)
        output_result(result)
        proceed = query_to_proceed()

# Tālāk seko definīcijas izmantotajām funkcijām

class fgcolors:
    RED = "\033[1;31m"
    GREEN = "\033[1;32m"
    NORMAL = "\033[0m"

def query_two_naturals() -> tuple[int,int]:
    """Vaicā un atgriež divus naturālus skaitļus."""
    a = -1 # a ir <= 0, kamēr tas nav korekti ievadīts
    while a <= 0:
        a_str = input("Lūdzu, ievadiet pirmo skaitli: ").strip()
        if a_str.isnumeric():
            a = int(a_str)
        if a<=0: # ja pareizi ievadīts, tad būs > 0
            print(fgcolors.RED+"Kļūda! Ievadītais skaitlis nav naturāls."+fgcolors.NORMAL)
    b = -1 # b ir <= 0, kamēr tas nav korekti ievadīts
    while b <= 0:
        b_str = input("Lūdzu, ievadiet otro skaitli: ").strip()
        if b_str.isnumeric():
            b = int(b_str)
        if b<=0: # ja pareizi ievadīts, tad būs > 0
            print(fgcolors.RED+"Kļūda! Ievadītais skaitlis nav naturāls."+fgcolors.NORMAL)
    return a, b

def get_divs(x: int) -> list:
    """Atgriež skaitļa naturālos dalītājus."""
    res = []
    i = 1
    while i*i<=x:
        if x%i==0:
            res.append(i)
            if((x//i)!=i):
                res.append(x//i)
        i += 1
    return res

def get_common_divisors(a: int, b: int):
    """Atgriež skaitļu kopīgos dalītājus."""
    res = []
    if b < a: # neliela veiktspējas, kas atkarīga no a dal. skaita, optimizācija
        a,b=b,a
    a_divs = get_divs(a)
    for d in a_divs:
        if b%d==0:
            res.append(d)
    res.sort()
    return res

def output_result(result: list):
    # fgcolors.GREEN iestata zaļu tekstu
    print(fgcolors.GREEN,"Kopīgie dalītāji: ",end="")
    print(" ".join(["-"+str(x) for x in result[::-1]]),end=" ")
    print(" ".join([str(x) for x in result]),end=" ")
    print(fgcolors.NORMAL) # atiestatam krāsu uz ierasto

def query_to_proceed() -> bool:
    """Vaicā lietotājam, vai viņš vēlas turpināt."""
    q_res = input("Vai turpināt (1), vai beigt (0)?\n")
    if q_res != "0" and q_res != "1":
        print(fgcolors.RED+"Kļūda. Tika gaidīta atbilde, kas ir 0 vai 1."+fgcolors.NORMAL)
        return query_to_proceed()
    return q_res=="1"

main()

File: a29/a29.cpp
/*
Autors: Krišjānis Petručeņa kp22084
Uzdevums: A29 - izvadīt divu naturālu skaitļu kopīgos dalītājus
Programma izveidota: 13.09.2022
*/

#include <iostream>
#include <vector>
#include <algorithm>
#include <stdexcept>

using namespace std;
using ll = long long;
using ull = unsigned long long;

void print_error(string err_msg)
{
    cout<<"\033[1;31m"<<err_msg<<"\033[0m"<<endl;
}

void print_success(string scc_msg)
{
    cout<<"\033[1;32m"<<scc_msg<<"\033[0m"<<endl;
}

/**
 * @brief Atgriež true, ja number satur tikai ciparus.
 */
bool is_natural(string number)
{
    bool has_non_zero = false; // naturals skaitlis satur ciparu, kas nav nulle
    for (char c : number)
    {
        if (c!='0') has_non_zero = true;
        if (c < '0' || c > '9')
            return false;
    }
    return has_non_zero&&true;
}

/**
 * @brief Vaicā lietotājam ievadīt divus naturālus skaitļus līdz (2^64)-1.
 * @return pair<ull,ull> - ievadītie naturālie skaitļi.
 */
pair<ull,ull> query_two_naturals()
{
    cout << "Lūdzu, ievadiet divus naturālus skaitļus!"<<endl;
    
    string a_str, b_str;
    cin >> a_str >> b_str;

    if (!is_natural(a_str) || !is_natural(b_str))
    {
        print_error("Kļūda! Ievadītais skaitlis nav naturāls.");
        return query_two_naturals();
    }

    try
    {
        ull a = stoull(a_str), b = stoull(b_str);
        return {a,b};
    }
    catch(out_of_range& oor)
    {
        print_error("Kļūda! Ievadītais skaitlis ir pārāk liels.");
        return query_two_naturals();
    }
}

/**
 * @brief Atgriež skaitļa naturālos dalītājus.
 * @param x naturāls skaitlis līdz (2^64)-1.
 */
vector<ull> get_divisors(ull x)
{
    vector<ull> divs; 
    for(ull i=1;i*i<=x;i++)
    {
        if(x%i==0)
        {
            divs.push_back(i);
            if((x/i)!=i)
                divs.push_back(x/i);
        }
    }
    return divs;
}

/**
 * @brief Atgriež kopīgos naturālos dalītājus augošā secībā.
 * @param a pirmais skaitlis līdz (2^64)-1.
 * @param b otrais skaitlis līdz (2^64)-1.
 */
vector<ull> get_common_divisors(ull a, ull b)
{

    if(b<a) swap(a,b); // neliela veiktspējas optimizācija

    vector<ull> a_divs = get_divisors(a);

    vector<ull> a_and_b_divs;
    for(ull x: a_divs)
        if(b%x==0)
            a_and_b_divs.push_back(x);

    sort(a_and_b_divs.begin(),a_and_b_divs.end());

    return a_and_b_divs;
}

/// Atgriež TRUE, ja lietotājs vēlas turpināt, un FALSE, ja nevēlas.
bool query_to_proceed()
{
    cout<<"Vai turpināt (1), vai beigt (0)?"<<endl;
    string response;
    cin>>response;
    if(response!="1"&&response!="0")
    {
        print_error("Kļūda! Tika sagaidīta atbilde, kas ir 0 vai 1.");
        return query_to_proceed();
    }
    return response=="1";
}

int main()
{

    cout<<"Autors: Krišjānis Petručeņa kp22084"<<endl;
    cout<<"Uzdevums: A29 - izvadīt divu naturālu skaitļu kopīgos dalītājus"<<endl;
    cout<<"Programma izveidota: 13.09.2022"<<endl;
    cout<<endl;
    
    bool proceed = true;
    do
    {
        // nepieciešams C++17 priekš structured binding
        auto [a,b] = query_two_naturals();

        vector<ull> result = get_common_divisors(a,b);

        vector<ull> rev_result = result; 
        reverse(rev_result.begin(),rev_result.end());// lai iegūtu negatīvos skaitļus augoši, apgriežam otrādāk pozitīvos

        string result_str = "Kopīgie dalītāji: ";
        // es neesmu pārliecināts par to, vai jāizvada negatīvie dalītāji
        for(ull x: rev_result)
            result_str += "-"+to_string(x)+" "; // - tiek likts kā simbols, jo x ir unsigned 
        for(ull x: result)
            result_str += to_string(x)+" ";

        print_success(result_str);

        proceed = query_to_proceed();
 
    } while (proceed);
}